### PR TITLE
fix: fetch tokens when loginInProgress is false

### PR DESCRIPTION
**Pull request by CodingDaria**
Looks like a multi tab scenario is not fully supported. Please check the scenario:

1. Call `logIn()` on your web page. Here: 
- `loginInProgress` is set to `true`;
- you are navigated away to your auth endpoint, don't log in.
2. Meanwhile, open another instance of your web page in a new tab. Here:
- `loginInProgress` is true but urlParams doesn't have `code`, so a "Bad authorization state" error is set and `clearStorage()` is called, `loginInProgress` is set to `false`;
3. Finish login at your first tab. You are redirected back to your page with an auth code, but `loginInProgress` is already false. Nothing happens, and code is not replaced from the URL. If `autoLogin` is true, you are redirected to the auth endpoint again.

**_My suggestion is to perform token fetch_** even if `loginInProgress` is false but is there is a code in URL.
Maybe we need more conditions?
I don't think this behavior is fixable outside of the package.

### CODE REVIEW

File: src/AuthContext.tsx [hunk 1]
```diff
@@ -192,6 +192,40 @@ export const AuthProvider = ({ authConfig, children }: IAuthProvider) => {
     )
   }
 
+  // This ref is used to make sure the 'fetchTokens' call is only made once.
+  // Multiple calls with the same code will, and should, return an error from the API
+  // See: https://beta.reactjs.org/learn/synchronizing-with-effects#how-to-handle-the-effect-firing-twice-in-development
+  const didFetchTokens = useRef(false)
+
+  function doFetchTokens(urlParams: URLSearchParams): void {
```
Comment (sebastianvitterso) [inline src/AuthContext.tsx:200]: Could just be called `fetchTokens`

**Review by sebastianvitterso — COMMENTED** [review]
Looks like it would work, and the idea isn't half bad. We had a discussion about multi-tab problems around a year ago, and figured that the current behavior was as good as we could get it without introducing a lot more complexity, but this seems to solve at least one problem with barely any more complexity, which is nice. 

I'll await @soofstad's thoughts on the matter.

File: src/AuthContext.tsx (hunk 1 above)
Comment (CodingDaria) [reply src/AuthContext.tsx:200]: > Could just be called `fetchTokens`

There is `fetchTokens` already...

Comment (sebastianvitterso) [reply src/AuthContext.tsx:200]: Fair point, so this is a wrapper of the fetching, which does more then. Hm.

`performTokenFetching`? 
`handleTokenFetch`? 
Something like that maybe. 

PS: Still waiting on @soofstad's feedback on the idea of the PR

**Comment by soofstad** [comment]
Hi @CodingDaria, thank you for the PR and for raising the issue.
I will make some time to look at this a bit closer, and test some stuff.
But the immediate feeling is that I'm not a huge fan. 
 - This is a quite narrow edge case 
 - Potential for an infinite redirect loop
 - A somewhat strange setup to have a variable for "loginInProgress", and then go ahead and ignore it
 - A breaking change. Any page with a queryParam called "code" will stop working

**Comment by soofstad** [comment]
Closing this for reasons state in above comment.
